h/viewderivers.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+
+def csp_protected_view(view, info):
+    """
+    A view deriver which adds Content-Security-Policy headers to responses.
+
+    By default, a global policy is applied to every view.
+
+    Individual views can opt out of CSP altogether by specifying a view option
+    ``csp_insecure_optout=True``. This is not recommended.
+    """
+    if not info.registry.settings.get('csp.enabled', False):
+        return view
+
+    policy = info.registry.settings.get('csp', {})
+    policy = "; ".join([
+        " ".join([k] + [v2 for v2 in v if v2 is not None])
+        for k, v in sorted(policy.items())
+        if [v2 for v2 in v if v2 is not None]
+    ])
+
+    if info.registry.settings.get('csp.report_only', False):
+        header_name = 'Content-Security-Policy-Report-Only'
+    else:
+        header_name = 'Content-Security-Policy'
+
+    # Views can set ``csp_insecure_optout=True`` in their view options to
+    # disable CSP for the view.
+    if info.options.get('csp_insecure_optout'):
```

Comment:
I think it would be clearer if this line was moved up to just after the global CSP setting check, so the function looks like:

```
if not csp_enabled_globally:
  return view

if not csp_enabled_for_view:
  return view

# Prepare and return the wrapper view
```